File: backend/src/taxonomy/taxonomy.ts
// The closed WorldSignal Taxonomy. Production tags MUST come from this list —
// the LLM is constrained to it, and the heuristic classifier matches keywords
// against the `keywords` field below.

export interface TaxonomyNode {
  code: string;
  label: string;
  description?: string;
  /** keywords used by the deterministic fallback classifier */
  keywords?: string[];
  aliases?: string[];
  children?: TaxonomyNode[];
}

export const TAXONOMY: TaxonomyNode[] = [
  {
    code: "POLITICS",
    label: "Politics",
    children: [
      { code: "POLITICS.ELECTIONS", label: "Elections", keywords: ["election", "ballot", "vote", "poll", "candidate", "campaign"] },
      { code: "POLITICS.POLICY", label: "Policy", keywords: ["policy", "legislation", "bill", "reform", "parliament", "congress"] },
      { code: "POLITICS.DIPLOMACY", label: "Diplomacy", keywords: ["diplomacy", "summit", "treaty", "sanction", "embassy", "foreign minister"] },
    ],
  },
  {
    code: "ECONOMY",
    label: "Economy",
    children: [
      { code: "ECONOMY.INFLATION", label: "Inflation", keywords: ["inflation", "cpi", "consumer price", "cost of living"] },
      { code: "ECONOMY.INTEREST_RATES", label: "Interest Rates", keywords: ["interest rate", "central bank", "rate hike", "basis points", "federal reserve", "rbi"] },
      { code: "ECONOMY.MARKETS", label: "Markets", keywords: ["stock market", "shares", "index", "nasdaq", "nifty", "sensex", "commodities", "crypto", "bitcoin"] },
      { code: "ECONOMY.JOBS", label: "Jobs & Employment", keywords: ["unemployment", "jobs report", "hiring", "labor market", "payroll"] },
    ],
  },
  {
    code: "BUSINESS",
    label: "Business",
    children: [
      { code: "BUSINESS.EARNINGS", label: "Earnings", keywords: ["earnings", "quarterly results", "revenue", "profit", "guidance"] },
      { code: "BUSINESS.MA", label: "Mergers & Acquisitions", keywords: ["acquisition", "merger", "takeover", "acquires", "buyout"] },
      { code: "BUSINESS.FUNDING", label: "Funding", keywords: ["funding", "series a", "series b", "raised", "venture", "valuation", "ipo"] },
      { code: "BUSINESS.LAYOFFS", label: "Layoffs", keywords: ["layoff", "job cuts", "restructuring", "redundancies"] },
    ],
  },
  {
    code: "TECHNOLOGY",
    label: "Technology",
    children: [
      { code: "TECHNOLOGY.AI", label: "Artificial Intelligence", keywords: ["artificial intelligence", "ai model", "machine learning", "llm", "openai", "chatbot", "neural"] },
      { code: "TECHNOLOGY.CYBERSECURITY", label: "Cybersecurity", keywords: ["data breach", "ransomware", "vulnerability", "hacked", "cyberattack", "malware", "cve"] },
      { code: "TECHNOLOGY.PRODUCT", label: "Product Launch", keywords: ["launches", "unveils", "release", "new device", "announced"] },
    ],
  },
  {
    code: "DISASTER",
    label: "Disaster",
    children: [
      { code: "DISASTER.EARTHQUAKE", label: "Earthquake", keywords: ["earthquake", "magnitude", "seismic", "tremor", "aftershock"] },
      { code: "DISASTER.FLOOD", label: "Flood", keywords: ["flood", "flooding", "inundation", "deluge"] },
      { code: "DISASTER.CYCLONE", label: "Cyclone / Storm", keywords: ["cyclone", "hurricane", "typhoon", "storm", "tornado"] },
      { code: "DISASTER.WILDFIRE", label: "Wildfire", keywords: ["wildfire", "bushfire", "forest fire", "blaze"] },
    ],
  },
  {
    code: "PUBLIC_HEALTH",
    label: "Public Health",
    children: [
      { code: "PUBLIC_HEALTH.OUTBREAK", label: "Disease Outbreak", keywords: ["outbreak", "epidemic", "pandemic", "virus", "infection", "cases surge"] },
      { code: "PUBLIC_HEALTH.DRUG", label: "Drug / Treatment", keywords: ["drug approval", "vaccine", "clinical trial", "fda approval", "treatment"] },
    ],
  },
  {
    code: "LEGAL",
    label: "Legal",
    children: [
      { code: "LEGAL.COURT_RULING", label: "Court Ruling", keywords: ["court", "ruling", "verdict", "supreme court", "judge", "lawsuit"] },
      { code: "LEGAL.REGULATION", label: "Regulation", keywords: ["regulation", "regulator", "compliance", "antitrust", "ban", "fine"] },
    ],
  },
  {
    code: "CONFLICT",
    label: "Conflict & Security",
    children: [
      { code: "CONFLICT.WAR", label: "War / Armed Conflict", keywords: ["war", "military", "airstrike", "troops", "ceasefire", "invasion", "missile"] },
      { code: "CONFLICT.TERRORISM", label: "Terrorism", keywords: ["terror", "bombing", "attack", "explosion", "shooting"] },
    ],
  },
  {
    code: "SPORTS",
    label: "Sports",
    children: [
      { code: "SPORTS.RESULT", label: "Match Result", keywords: ["wins", "defeats", "final", "tournament", "championship", "match"] },
      { code: "SPORTS.TRANSFER", label: "Transfer", keywords: ["transfer", "signs", "signing", "contract", "deal"] },
    ],
  },
  {
    code: "GENERAL",
    label: "General",
    children: [{ code: "GENERAL.OTHER", label: "Other / Uncategorized", keywords: [] }],
  },
];

/** Flattened list of every node (domains + leaves). */
export function flattenTaxonomy(nodes: TaxonomyNode[] = TAXONOMY): TaxonomyNode[] {
  const out: TaxonomyNode[] = [];
  for (const n of nodes) {
    out.push(n);
    if (n.children) out.push(...flattenTaxonomy(n.children));
  }
  return out;
}

/** Only leaf nodes (the assignable tags). */
export function leafTags(): TaxonomyNode[] {
  return flattenTaxonomy().filter((n) => !n.children);
}

/** The set of valid codes — used to validate LLM output. */
export const VALID_CODES: Set<string> = new Set(flattenTaxonomy().map((n) => n.code));

export const FALLBACK_CODE = "GENERAL.OTHER";

File: backend-go/internal/taxonomy/taxonomy.go
// Package taxonomy ports backend/src/taxonomy/taxonomy.ts. The closed vocabulary
// the LLM and heuristic classifier are constrained to. JSON serialization is
// byte-compatible with the TS object-literal order: domains emit {code,label,
// children}; leaves emit {code,label,keywords} (keywords present even when empty).
package taxonomy

import (
	"bytes"

	"github.com/worldsignal/backend/internal/jsonx"
)

// Node is a taxonomy entry. Keywords is non-nil exactly for leaf nodes (matching
// the TS data where every leaf sets `keywords` and domains never do); Children is
// non-nil for domains.
type Node struct {
	Code     string
	Label    string
	Keywords []string
	Children []Node
}

// MarshalJSON emits keys in the fixed order code,label,keywords?,children? — a
// node never has both keywords and children, so this matches both TS shapes.
func (n Node) MarshalJSON() ([]byte, error) {
	var b bytes.Buffer
	b.WriteByte('{')
	code, _ := jsonx.Marshal(n.Code)
	label, _ := jsonx.Marshal(n.Label)
	b.WriteString(`"code":`)
	b.Write(code)
	b.WriteString(`,"label":`)
	b.Write(label)
	if n.Keywords != nil {
		kw, err := jsonx.Marshal(n.Keywords)
		if err != nil {
			return nil, err
		}
		b.WriteString(`,"keywords":`)
		b.Write(kw)
	}
	if n.Children != nil {
		ch, err := jsonx.Marshal(n.Children)
		if err != nil {
			return nil, err
		}
		b.WriteString(`,"children":`)
		b.Write(ch)
	}
	b.WriteByte('}')
	return b.Bytes(), nil
}

func leaf(code, label string, keywords ...string) Node {
	if keywords == nil {
		keywords = []string{}
	}
	return Node{Code: code, Label: label, Keywords: keywords}
}

func domain(code, label string, children ...Node) Node {
	return Node{Code: code, Label: label, Children: children}
}

// Taxonomy is the closed WorldSignal taxonomy tree.
var Taxonomy = []Node{
	domain("POLITICS", "Politics",
		leaf("POLITICS.ELECTIONS", "Elections", "election", "ballot", "vote", "poll", "candidate", "campaign"),
		leaf("POLITICS.POLICY", "Policy", "policy", "legislation", "bill", "reform", "parliament", "congress"),
		leaf("POLITICS.DIPLOMACY", "Diplomacy", "diplomacy", "summit", "treaty", "sanction", "embassy", "foreign minister"),
	),
	domain("ECONOMY", "Economy",
		leaf("ECONOMY.INFLATION", "Inflation", "inflation", "cpi", "consumer price", "cost of living"),
		leaf("ECONOMY.INTEREST_RATES", "Interest Rates", "interest rate", "central bank", "rate hike", "basis points", "federal reserve", "rbi"),
		leaf("ECONOMY.MARKETS", "Markets", "stock market", "shares", "index", "nasdaq", "nifty", "sensex", "commodities", "crypto", "bitcoin"),
		leaf("ECONOMY.JOBS", "Jobs & Employment", "unemployment", "jobs report", "hiring", "labor market", "payroll"),
	),
	domain("BUSINESS", "Business",
		leaf("BUSINESS.EARNINGS", "Earnings", "earnings", "quarterly results", "revenue", "profit", "guidance"),
		leaf("BUSINESS.MA", "Mergers & Acquisitions", "acquisition", "merger", "takeover", "acquires", "buyout"),
		leaf("BUSINESS.FUNDING", "Funding", "funding", "series a", "series b", "raised", "venture", "valuation", "ipo"),
		leaf("BUSINESS.LAYOFFS", "Layoffs", "layoff", "job cuts", "restructuring", "redundancies"),
	),
	domain("TECHNOLOGY", "Technology",
		leaf("TECHNOLOGY.AI", "Artificial Intelligence", "artificial intelligence", "ai model", "machine learning", "llm", "openai", "chatbot", "neural"),
		leaf("TECHNOLOGY.CYBERSECURITY", "Cybersecurity", "data breach", "ransomware", "vulnerability", "hacked", "cyberattack", "malware", "cve"),
		leaf("TECHNOLOGY.PRODUCT", "Product Launch", "launches", "unveils", "release", "new device", "announced"),
	),
	domain("DISASTER", "Disaster",
		leaf("DISASTER.EARTHQUAKE", "Earthquake", "earthquake", "magnitude", "seismic", "tremor", "aftershock"),
		leaf("DISASTER.FLOOD", "Flood", "flood", "flooding", "inundation", "deluge"),
		leaf("DISASTER.CYCLONE", "Cyclone / Storm", "cyclone", "hurricane", "typhoon", "storm", "tornado"),
		leaf("DISASTER.WILDFIRE", "Wildfire", "wildfire", "bushfire", "forest fire", "blaze"),
	),
	domain("PUBLIC_HEALTH", "Public Health",
		leaf("PUBLIC_HEALTH.OUTBREAK", "Disease Outbreak", "outbreak", "epidemic", "pandemic", "virus", "infection", "cases surge"),
		leaf("PUBLIC_HEALTH.DRUG", "Drug / Treatment", "drug approval", "vaccine", "clinical trial", "fda approval", "treatment"),
	),
	domain("LEGAL", "Legal",
		leaf("LEGAL.COURT_RULING", "Court Ruling", "court", "ruling", "verdict", "supreme court", "judge", "lawsuit"),
		leaf("LEGAL.REGULATION", "Regulation", "regulation", "regulator", "compliance", "antitrust", "ban", "fine"),
	),
	domain("CONFLICT", "Conflict & Security",
		leaf("CONFLICT.WAR", "War / Armed Conflict", "war", "military", "airstrike", "troops", "ceasefire", "invasion", "missile"),
		leaf("CONFLICT.TERRORISM", "Terrorism", "terror", "bombing", "attack", "explosion", "shooting"),
	),
	domain("SPORTS", "Sports",
		leaf("SPORTS.RESULT", "Match Result", "wins", "defeats", "final", "tournament", "championship", "match"),
		leaf("SPORTS.TRANSFER", "Transfer", "transfer", "signs", "signing", "contract", "deal"),
	),
	domain("GENERAL", "General",
		leaf("GENERAL.OTHER", "Other / Uncategorized"),
	),
}

// FallbackCode is used when nothing else matches.
const FallbackCode = "GENERAL.OTHER"

// Flatten returns every node (domains + leaves), depth-first, matching flattenTaxonomy.
func Flatten(nodes []Node) []Node {
	var out []Node
	for _, n := range nodes {
		out = append(out, n)
		if n.Children != nil {
			out = append(out, Flatten(n.Children)...)
		}
	}
	return out
}

// LeafTags returns only assignable leaf nodes (no children).
func LeafTags() []Node {
	var out []Node
	for _, n := range Flatten(Taxonomy) {
		if n.Children == nil {
			out = append(out, n)
		}
	}
	return out
}

// ValidCodes is the set of every code (domains + leaves), used to validate LLM output.
var ValidCodes = func() map[string]struct{} {
	m := map[string]struct{}{}
	for _, n := range Flatten(Taxonomy) {
		m[n.Code] = struct{}{}
	}
	return m
}()
